exonum/src/events/noise/wrapper.rs

```diff
@@ -126,35 +126,50 @@ impl NoiseWrapper {
     /// 4. Append all encrypted packets in corresponding order.
     /// 5. Write result message to `buf`
     pub fn encrypt_msg(&mut self, msg: &[u8], buf: &mut BytesMut) -> Result<Option<()>, io::Error> {
-        let mut len = 0_usize;
-        let mut encoded_message = vec![0_u8; 0];
+        let len = self.encrypted_msg_len(msg.len());
+        let mut encoded_message = Vec::with_capacity(len);
 
-        msg.chunks(NOISE_MAX_MESSAGE_LENGTH - TAG_LENGTH)
-            .for_each(|msg| {
-                let (written_bytes, written) = self.write(msg).unwrap();
-                encoded_message.extend_from_slice(&written);
-                len += written_bytes;
-            });
+        for msg in msg.chunks(MAX_MESSAGE_LENGTH - TAG_LENGTH) {
+            let written = self.write(msg)?;
+            encoded_message.extend_from_slice(&written);
+        }
 
         let mut msg_len_buf = vec![0_u8; NOISE_HEADER_LENGTH];
 
         LittleEndian::write_u32(&mut msg_len_buf, len as u32);
         let encoded_message = &encoded_message[0..len];
         msg_len_buf.extend_from_slice(encoded_message);
         buf.extend_from_slice(&msg_len_buf);
+
+        debug_assert_eq!(len, encoded_message.len());
         Ok(None)
     }
 
-    fn read(&mut self, input: &[u8], len: usize) -> Result<(usize, Vec<u8>), NoiseError> {
+    fn read(&mut self, input: &[u8], len: usize) -> Result<Vec<u8>, NoiseError> {
```

Comment:
From the performance perspective, the most right way to do things would probably be to organize read as

```rust
fn read(&mut self, input: &[u8], output: &mut [u8]) -> Result<usize, NoiseError>
```

in order to be able to reuse the output buffer in `decrypt_msg`. At this point, it would probably be easier just to inline `self.session.read_message` :upside_down_face: